### PR TITLE
Add Nonce functionality to solana-web3

### PR DESCRIPTION
**Pull request by CriesofCarrots**
Depends on https://github.com/solana-labs/solana/pull/7645 

Closes https://github.com/solana-labs/solana/issues/7583
and addresses web3 concern of https://github.com/solana-labs/solana/issues/7700

### CODE REVIEW

**Comment by CriesofCarrots** [comment]
@mvines , the most controversial part of this is commit 0dd4c89 that adds a `nonceInfo` field to Transaction, and trumps `recentBlockhash` in `Transaction._getSignData()` and `Connection.sendTransaction()`. Are you okay with this approach?

I would have liked to only store the Nonce account pubkey and the Nonce hash in that object, but building the NonceAdvance instruction dynamically in `_getSignData()` caused a circular dependency issue between Transaction and SystemProgram that did not seem simpler when untangled.

File: src/connection.js [hunk 1]
```diff
@@ -1149,6 +1150,56 @@ export class Connection {
     };
   }
 
+  /**
+   * Fetch a list of Transactions and transaction statuses from the cluster
+   * for a confirmed block
+   */
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply src/connection.js]: 🤦‍♀

File: src/connection.js [hunk 2]
```diff
@@ -1149,6 +1150,55 @@ export class Connection {
...
+    if (res.error) {
+      throw new Error(res.error.message);
+    }
+    assert(typeof res.result !== 'undefined');
+
+    const isV021 =
+      typeof res.result.context !== 'undefined' &&
+      typeof res.result.value !== 'undefined';
+
+    const slot = isV021 ? res.result.context.slot : NaN;
```
Comment (mvines) [inline src/connection.js]: I think we can drop v0.21 support anytime now

File: src/connection.js [hunk 3]
```diff
@@ -1149,6 +1150,55 @@ export class Connection {
...
+    const value = NonceAccount.fromAccountData(Buffer.from(resultValue.data));
+
+    return {
+      context: {
+        slot,
+      },
+      value,
+    };
+  }
+  async getNonce(
```
Comment (mvines) [inline src/connection.js]: whitespace before new function, and esdoc header please

**Comment by mvines** [comment]
> Are you okay with this approach?

Yep seems reasonable!

File: src/connection.js (hunk 2 above)
Comment (CriesofCarrots) [reply src/connection.js]: Sounds good. I'll do that in a separate PR.

File: src/connection.js (hunk 3 above)
Comment (CriesofCarrots) [reply src/connection.js]: I was following the pattern of all the other `AndContext` method pairs on this.
I'll fix this up for all methods it that aforementioned separate PR.

**Pull request merged**